# In this notebook we're going to cover some of the most fundaental concepts of tensors using Tensorflow

More specifically, we're going to cover:
* Introduction to tensors
* Getting information from tensors
* Manipulating tensors
* Tensors & NumPy
* Using @tf.function (a way to speed up your regular Python functions)
* Exercoses to try for yourself!

## Introduction to Tensors

In [6]:
pip install tensorflow

Note: you may need to restart the kernel to use updated packages.


  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
You should consider upgrading via the 'c:\Users\capma\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


In [1]:
# Import TensorFlow
import tensorflow as tf
print(tf.__version__)

2.18.0


In [ ]:
# Create tensors with tf.constant()
scalar = tf.constant(7) # creates a constant tensor with the value 7
scalar

<tf.Tensor: shape=(), dtype=int32, numpy=7>

In [10]:
# Check the number of dimensions of a tensor (ndim stands for number of dimensions)
scalar.ndim

0

In [12]:
# Create a vector
vector = tf.constant([10,10])
vector

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([10, 10], dtype=int32)>

In [13]:
# Check the dimension of our vector
vector.ndim

1

In [15]:
# Create a matrix (has more than 1 dimension)
matrix = tf.constant([[10,7],
                      [7,10]])
matrix

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 7, 10]], dtype=int32)>

In [16]:
matrix.ndim

2

`ndim` returns us the dimension our arrays are in

In [17]:
# Create another matrix
another_matrix = tf.constant([[10.,7.],
                              [3.,2.,],
                              [8.,9.,]], dtype = tf.float16) # specify the data type with dtype parameters

another_matrix

<tf.Tensor: shape=(3, 2), dtype=float16, numpy=
array([[10.,  7.],
       [ 3.,  2.],
       [ 8.,  9.]], dtype=float16)>

so we have `float16` and `int32`. The number after the word means the number of bits it takes up to store the number. Lower bits means lower space to take up to store a value.

`16-bit` - values from -65,504 to 65,504, particularly used for deep learning and fast computation due to lower memory usage.

`32-bit` - values from -2,147,483,648 to 2,147,483,647, common for counting and inexing/categorizing data.

In [20]:
# What's the number dimensions of another_matrix
another_matrix.ndim

2

In [21]:
# Let's create a tensor
tensor = tf.constant([[[1,2,3,],
                       [4,5,6]],
                       [[7,8,9],
                        [10,11,12]],
                        [[13,14,15],
                         [16,17,18]]])

tensor

<tf.Tensor: shape=(3, 2, 3), dtype=int32, numpy=
array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18]]], dtype=int32)>

In [22]:
tensor.ndim

3

What we've created so far:
* Scalar: a single number
* Vector: a number with direction (e.g. wind speed and direction)
* Matrix: a 2-dimensional array of numbers
* Tensor: an n-dimensional aray of numbers (where n can be any number, a 0-dimensional tensor is a scalar, a 1 dimensional tensor is a vector)

### Creating tensors with `tf.Variable`

In [24]:
# Create the same tensor with tf.Variable() as above
changeable_tensor = tf.Variable([10,7])
unchangeable_tensor = tf.constant([10,7])
changeable_tensor, unchangeable_tensor

(<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([10,  7], dtype=int32)>,
 <tf.Tensor: shape=(2,), dtype=int32, numpy=array([10,  7], dtype=int32)>)

In [25]:
# Let's try change one of the elements in our changeable tensor
changeable_tensor[0] = 7
changeable_tensor

TypeError: 'ResourceVariable' object does not support item assignment

In [27]:
# How about we try .assign()
changeable_tensor[0].assign(7)
changeable_tensor

<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([7, 7], dtype=int32)>

In [28]:
# Let's try change our unchangeable tensor
unchangeable_tensor[0].assign(7)
unchangeable_tensor

AttributeError: 'tensorflow.python.framework.ops.EagerTensor' object has no attribute 'assign'

How should you choose between `constant` or `variable`?

That depends on your problem, though most of the time, TensorFlow automatically picks one for you. (When loading/modelling data)

### Creating random tensors

Random tensors, are arbitrary values that contain numbers. But why would we want randomized tensors?

This is to initialize weights (patterns), where the neural networks use as a starting point to learn the data.

They take the randomized data, and start refining them bit by bit, as they learn and understand the patterns of the data.

#### How a network learns:
![](00-how-a-network-learns.png)

*Networks learn from random patterns (1) then going through provided data (2) whilt updating its randomized patterns, to represent provided data*

We can create random tensors by using the `tf.random.Generator` class.

In [4]:
# Create two random, but same tensors
random_1 = tf.random.Generator.from_seed(42) # set seed for reproducibility
random_1 = random_1.normal(shape=(3,2)) # create tensor from a normal distribution
random_2 = tf.random.Generator.from_seed(42)
random_2 = random_2.normal(shape=(3,2))

# Are they equal?
random_1, random_2, random_1 == random_2

(<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702],
        [ 0.07595026, -1.2573844 ],
        [-0.23193763, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702],
        [ 0.07595026, -1.2573844 ],
        [-0.23193763, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=bool, numpy=
 array([[ True,  True],
        [ True,  True],
        [ True,  True]])>)

Its not truly random, but pseudorandom. The randomness of numbers are controlled by the `seed`. This allows consistency in random numbers when we need it, such as comparing 2 different ML models.

So setting the same `seed` for the data, will give us the same set of randomized number.

In [5]:
# Create two random and different tensors
random_3 = tf.random.Generator.from_seed(42)
random_3 = random_3.normal(shape=(3,2))
random_4 = tf.random.Generator.from_seed(11)
random_4 = random_4.normal(shape=(3,2))

# Check tensors if equal
random_3, random_4, random_3 == random_4

(<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702],
        [ 0.07595026, -1.2573844 ],
        [-0.23193763, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[ 0.27305737, -0.29925638],
        [-0.3652325 ,  0.61883307],
        [-1.0130816 ,  0.28291714]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=bool, numpy=
 array([[False, False],
        [False, False],
        [False, False]])>)

How about shuffling the order of tensors?

This is useful, for example a case scenario, where you have 15,000 images of cats and dogs. However, the first 10,000 photos are cats, and the last 5,000 are dogs. The order of the images may affect the neural network's learning, as it might put too much focus on order of images, and predicting the front photos as cats.

So sometimes its good to shuffle things around.

In [9]:
# Shuffling a tensor
not_shuffled = tf.constant([[10,7],
                            [3,4],
                            [2,5]])
# Getting a different result every time you run this code
tf.random.shuffle(not_shuffled)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 2,  5],
       [ 3,  4]], dtype=int32)>

In [ ]:
# Shuffle randomly but in the same order, through seed (but won't actually be the same)
tf.random.shuffle(not_shuffled, seed=42)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

Why didn't the seed apply?

Because of the 4th rule in set_seed() documentation. `tf.random.set_seed(42)` sets the global seed, and therefore the seed parameter can be used in `tf.random.shuffle(seed=42)`.

"Operations relying on random seeds are driven from two seeds: the global and operation-level seeds."

In [15]:
# Shuffle in same order, set global random seed
tf.random.set_seed(42)

# Set the operation random seed
tf.random.shuffle(not_shuffled, seed=42)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

Not as often used, you can create `tf.ones()` to create tensors full of ones, and `tf.zeros()` for tensors full of zeros.

In [16]:
tf.ones(shape=(3,2))

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[1., 1.],
       [1., 1.],
       [1., 1.]], dtype=float32)>

In [23]:
tf.zeros(shape=(3,2))

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[0., 0.],
       [0., 0.],
       [0., 0.]], dtype=float32)>

NumPy arrays are convertible to tensors.

NOTE:
Main difference between tensors and NumPy arrays is that tensors can use GPU, making neural network processing much faster

In [16]:
import numpy as np
numpy_A= np.arange(1, 25, dtype=np.int32) # create NumPy array from 1 to 25, exclusive
A = tf.constant(numpy_A,
                shape=[2,4,3]) # note the shape total (2*4*3) has to match number of elements in the array
numpy_A, A

(array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23, 24], dtype=int32),
 <tf.Tensor: shape=(2, 4, 3), dtype=int32, numpy=
 array([[[ 1,  2,  3],
         [ 4,  5,  6],
         [ 7,  8,  9],
         [10, 11, 12]],
 
        [[13, 14, 15],
         [16, 17, 18],
         [19, 20, 21],
         [22, 23, 24]]], dtype=int32)>)

### Getting information from tensors (shape, rank, size)

There are times where you want speciic pieces of infomations form your tensors, such as:
* **Shape**: The length (number of elements) of each of the dimensions of a tensor
* **Rank**: The number of tensor dimensions. Scalar = 0, Vector = 1, Matrix = 2, Tensor = x (any amount)
* **Axis** or **Dimension**: A particular dimension of a tensor.
* **Size**: The total number of items in the tensor.

These are used, when lining up the shape of the data to the shape of your model. Such as making sure the shape of your image tensors, are the same shape as your model's input layer.

Mismatch in shape causes errors, or inefficient processing in the neural networks

We've seen the use of `ndim` attribute, let's see the rest.

In [26]:
# Create a rank 4 tensor (4 dimensions)
rank_4_tensor = tf.zeros([2,3,4,5])
rank_4_tensor

<tf.Tensor: shape=(2, 3, 4, 5), dtype=float32, numpy=
array([[[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]],


       [[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]]], dtype=float32)>

In [28]:
rank_4_tensor.shape, rank_4_tensor.ndim, tf.size(rank_4_tensor)

(TensorShape([2, 3, 4, 5]), 4, <tf.Tensor: shape=(), dtype=int32, numpy=120>)

In [30]:
# Get various attributes of tensor
print('Datatype of every element:', rank_4_tensor.dtype)
print('Number of dimensions (rank):', rank_4_tensor.ndim)
print('Shape of tensor:', rank_4_tensor.shape)
print('Elements along axis(index) 0 of tensor:', rank_4_tensor.shape[0])
print('Elements along last axis(index) of tensor:', rank_4_tensor.shape[-1])
print('Total number of elements (2*3*4*5)', tf.size(rank_4_tensor).numpy()) # .numpy() converts to NumPy array

Datatype of every element: <dtype: 'float32'>
Number of dimensions (rank): 4
Shape of tensor: (2, 3, 4, 5)
Elements along axis(index) 0 of tensor: 2
Elements along last axis(index) of tensor: 5
Total number of elements (2*3*4*5) 120


You can also index tensors just like Python lists.

In [33]:
# Get the first 2 items of each dimension
rank_4_tensor[:2, :2, :2, :2] # retriees from index 0 to 1, excludes 2

<tf.Tensor: shape=(2, 2, 2, 2), dtype=float32, numpy=
array([[[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]],


       [[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]]], dtype=float32)>

In [34]:
# get the dimension from each index except for the final one
rank_4_tensor[:1,:1,:1,:]

<tf.Tensor: shape=(1, 1, 1, 5), dtype=float32, numpy=array([[[[0., 0., 0., 0., 0.]]]], dtype=float32)>

In [35]:
# Create a rank 2 tensor (2 dimensions)
rank_2_tensor = tf.constant([[10,7],
                             [3,4]])

# Get the last item of each row
rank_2_tensor[:,:-1]

<tf.Tensor: shape=(2, 1), dtype=int32, numpy=
array([[10],
       [ 3]], dtype=int32)>

If you want to add dimensions, whilst keeping the same data present, use `tf.newaxis`.

In [36]:
# Add an extra dimension (to the end)
rank_3_tensor = rank_2_tensor[..., tf.newaxis] # '...' > in Python means 'all dimensions prior to'
rank_2_tensor, rank_3_tensor

(<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
 array([[10,  7],
        [ 3,  4]], dtype=int32)>,
 <tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
 array([[[10],
         [ 7]],
 
        [[ 3],
         [ 4]]], dtype=int32)>)

Our `tf.newaxis` makes a new dimension, past all the other present dimensions, due to the rule `...`

You can achieve the same using `tf.expand_dims()`.

In [38]:
tf.expand_dims(rank_2_tensor, axis=-1) # adding it at last row with axis/index -1

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]], dtype=int32)>

### Manipulating tensors (tensor operations)

Finding patterns in tensors (numerical representation of data) requires manipulaing them.

Again, when building models in TensorFlow, much of this pattern discovery is done for you.


### Basic operations

You can perform basic mathematical operations like + , -, * directly onto the tensors.

In [39]:
# You can add values to a tensor uding the addition operator
tensor = tf.constant([[10,7], [3,4]])
tensor + 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[20, 17],
       [13, 14]], dtype=int32)>

Since we used `tf.constant()`, the original tensor is unchanged. The addition is done on a separate copy

In [40]:
# Original tensor unchanged
tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

We can try other operators as well

In [41]:
tensor * 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[100,  70],
       [ 30,  40]], dtype=int32)>

In [42]:
tensor - 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 0, -3],
       [-7, -6]], dtype=int32)>

In [43]:
tensor / 2

<tf.Tensor: shape=(2, 2), dtype=float64, numpy=
array([[5. , 3.5],
       [1.5, 2. ]])>

There is an equivalent to these operators, called TensorFlow function. Use it wherever possible, as it performs much quicker than the operators above.

In [44]:
# Use tensorflow function, equivalent to *
tf.multiply(tensor, 10)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[100,  70],
       [ 30,  40]], dtype=int32)>

In [45]:
# the original tensor is still unchanged
tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

### Matrix multiplication

A common operation thats used is the `matrix multiplication`. Its implemented with `tf.matmul()`.

There two main rules to be aware of:

1. The inner dimensions must match:
* (3,5) @ (3.5) won't work
* (5,3) @ (3,5) will work
* (3,5) @ (5,3) will work

2. The resulting matrix has the shape of the outer dimensions:
* (5,3) @ (3,7) -> (5,7)
* (3,5) @ (5,3) -> (3,3)


In [47]:
# Matrix multiplication in TensorFlow
print(tensor)
tf.matmul(tensor, tensor)

tf.Tensor(
[[10  7]
 [ 3  4]], shape=(2, 2), dtype=int32)


<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]], dtype=int32)>

In [48]:
# Matrix multiplication with pythn operator '@'
tensor @ tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]], dtype=int32)>

But wait, it's not multiplying in the way I was expecting. 10*10 should be 100 right? Why is there 121?

There's an explanation to the way they multiply two different tensors:
![](matrix-multiplication.png)

we got our two tensors, 

10, 7 | 10, 7

 3, 4 | 3, 4

turn tensor 1, 90 degrees clockwise

3, 10 | 10, 7

 4, 7 | 3, 4

 then multiply the inner number together: *10 * 10 , 7 * 3*
 
 then the inner numbers from tensor 1, to outer tensor 2: *10 * 7 , 7 * 4*

 then do the other side of tensor 1, with inner tensor 2: *3 * 10 , 4 * 3*

 and again to the outer tensor 2: *3 * 7 , 4 * 4*

 Here's another graphic from the course, in a more appealing way:
 ![](00-lining-up-dot-products.png)

Now, let's create tensors that are mismatched shapes, unlike the one above

In [49]:
# Create (3,2) tensor
X = tf.constant([[1,2],
                 [3,4],
                 [5,6]])

# Another (3,2) tensor
Y = tf.constant([[7,8],
                 [9,10],
                 [11,12]])

X, Y

(<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
 array([[1, 2],
        [3, 4],
        [5, 6]], dtype=int32)>,
 <tf.Tensor: shape=(3, 2), dtype=int32, numpy=
 array([[ 7,  8],
        [ 9, 10],
        [11, 12]], dtype=int32)>)

In [50]:
# Try to matrix multiply them
X @ Y

InvalidArgumentError: {{function_node __wrapped__MatMul_device_/job:localhost/replica:0/task:0/device:CPU:0}} Matrix size-incompatible: In[0]: [3,2], In[1]: [3,2] [Op:MatMul] name: 

Why doesn't it work, well because the inner dimensions don't match each other. Since we've learnt how matrixes multiply, it helps us understand why they won't work with each other.

1, 2 | 7, 8

3, 4 | 9, 10

5, 6 | 11, 12

Now let's perform the 90 degree clockwise flip method on tensor 1.

5, 3, 1 | 7, 8

6, 4, 2 | 9, 10

| 11, 12

We have a problem where we dont have values for the 3rd row to multiply with!

Now we know why it doesn't work, we need to reshape it, so it fits the rules:
* `tf.reshape()` - allows us to reshape a tensor into a defined shape
* `tf.transpose()` - switches the dimensions of a given tensor

In [51]:
# Example of reshape (3,2) > (2,3)
tf.reshape(Y, shape=(2,3))

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[ 7,  8,  9],
       [10, 11, 12]], dtype=int32)>

In [53]:
# Try matrix multiplication with reshaped Y
X @ tf.reshape(Y, shape=(2,3))

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 27,  30,  33],
       [ 61,  68,  75],
       [ 95, 106, 117]], dtype=int32)>

Let's try it with X instead, and using `tf.transpose()` and `tf.matmul()`.

In [54]:
# Example of transpose (3,2) > (2,3)
tf.transpose(X)

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[1, 3, 5],
       [2, 4, 6]], dtype=int32)>

In [55]:
# Try matrix multiplification
tf.matmul(tf.transpose(X), Y)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 89,  98],
       [116, 128]], dtype=int32)>

In [56]:
# You can do the sae with parameters/booleans
tf.matmul(a=X, b=Y, transpose_a=True, transpose_b=False)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 89,  98],
       [116, 128]], dtype=int32)>

Notice how our transposition/reshaping of `X` and `Y` changes the results? It's because of rule 1, the inner values must be the same

* `(3,2) @ (2,3)` > `(3,3)` due to reshaping `Y`
* `(2,3) @ (3,2)` > `(2,2)` due to reshaping `X`

Data manipulation reminder:
You'll spend lots of time in machine learning, working with neural networks, reshaping data and preparing it for various operations.

### The dot product

Multiplying matrices with each over, can also be referred as the dot product.

You can perform `tf.matmul()` operation using `tf.tensordot()`

In [57]:
# Perform the dot product on X and Y (we'll transpose X so it won't cause errors)
tf.tensordot(tf.transpose(X), Y, axes=1)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 89,  98],
       [116, 128]], dtype=int32)>

Despite `reshape` and `transpose` do very similar jobs, the output when you multiply the matrices is different, after performing these two functions.

Let's see these in action.

In [60]:
# Perform matrix multiplication between X and Y (transposed)
tf.matmul(X, tf.transpose(Y))

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 23,  29,  35],
       [ 53,  67,  81],
       [ 83, 105, 127]], dtype=int32)>

In [61]:
# Perform matrix multiplication between X and Y (reshaped)
tf.matmul(X, tf.reshape(Y, (2,3)))

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 27,  30,  33],
       [ 61,  68,  75],
       [ 95, 106, 117]], dtype=int32)>

Theoretically the reshaping should result in the same outcome, as we're only reshaping only one of the tensors. Something must be going on between the two functions.

In [62]:
# Check shaped of Y, reshaped Y and transposed Y
Y.shape, tf.reshape(Y, (2,3)).shape, tf.transpose(Y).shape

(TensorShape([3, 2]), TensorShape([2, 3]), TensorShape([2, 3]))

Sure we got the same shape between the two functions. How about the values within Y after its been reshaped?

In [64]:
# Check values of Y, reshape Y and transposed Y
print('Normal Y:')
print(Y, '/n') # '/n' means new line

print('Y reshaped to (2,3):')
print(tf.reshape(Y, (2,3)), '/n')

print('Y transposed:')
print(tf.transpose(Y))

Normal Y:
tf.Tensor(
[[ 7  8]
 [ 9 10]
 [11 12]], shape=(3, 2), dtype=int32) /n
Y reshaped to (2,3):
tf.Tensor(
[[ 7  8  9]
 [10 11 12]], shape=(2, 3), dtype=int32) /n
Y transposed:
tf.Tensor(
[[ 7  9 11]
 [ 8 10 12]], shape=(2, 3), dtype=int32)


Despite producing the same shaped, they organize the values within them differently. This can be explained by their method of choice:

* `tf.reshape()` - change the shape of the given tensor first, then inserting the values in order (from lowest, to highest index, and continuing for the next set in other dimensions. [[7,8], [9,10], [11,12]])

* `tf.tranpose()` - swap the order of the axes, aka performing a 90degree clockwise shift of the values.

So which to use?

The thing is, these operations are automatically chosen for you while training your neural networks. But generally, its up to you on which function you'd like to use.

### Matrix multiplication tidbits

* If we transposed `Y`, it would be represented as `Y^T` (^T represents transpose).

* There's a hands-on demo of matrix multiplication: [http://matrixmultiplication.xyz/](http://matrixmultiplication.xyz/)

![](00-matrix-multiply-crop.gif)

Again it shows us how we flip 90 degrees on one tensor. For tensor 2, you counterclockwise the flip, and get your results in ordern from left to right, top to bottom.

### Changing the datatype of a tensor

Sometimes, you want to alter the default datatype of your tensor.

This is common if you want to compute something with less precision, say a 16-bit float, rather than the default 32-bit float.

This is useful if computer capacity for you is a luxury, such as on old pcs or phones.

You can change datatype of tensor with `tf.cast()`

In [1]:
# Create a new tensor with default datatype (float32)
B = tf.constant([1.7,7.4])

# Create a new tensor with default datatype (int32)
C = tf.constant([1,7])
B, C

NameError: name 'tf' is not defined

In [67]:
# Change from float32 to float16
B = tf.cast(B, dtype=tf.float16)
B


<tf.Tensor: shape=(2,), dtype=float16, numpy=array([1.7, 7.4], dtype=float16)>

In [68]:
# Change form int32 to float32
C = tf.cast(C, dtype=tf.float32)
C

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([1., 7.], dtype=float32)>

### Getting absolute values

`tf.abs()` gets all the values to default positive.

In [69]:
# Create tensor with negative values
D = tf.constant([-7,-10])
D

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ -7, -10], dtype=int32)>

In [70]:
# Get the absolute values
tf.abs(D)

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ 7, 10], dtype=int32)>

### Finding the min, max, mean, sum (aggregation)
You can quickly aggregate (perform calculations on the whole tensor) the tensor, to find things like min, max, mean, sum, etc.

To perform those actions, the methods have this syntax: `reduce()_[action]`, such as:
* `tf.reduce_min()`
* `tf.reduce_max()`
* `tf.reduce_mean()`
* `tf.reduce_sum()`

Their method name's should be pretty self explanatory on what they perform

**NOTE**: They are typically labeled under the `math` module, e.g. `tf.math.reduce_min()`, but can still be used under its alias, `tf.reduce_min()`.

In [71]:
# Create a tensor with 50 random values between 0 and 100
E = tf.constant(np.random.randint(low=0, high=100, size=50))
E

<tf.Tensor: shape=(50,), dtype=int32, numpy=
array([47, 98, 72, 18, 67, 29, 57, 67, 22, 36, 93, 46, 88, 92, 46, 10, 68,
       35, 55, 10, 73, 37, 20, 75, 10, 35,  9, 78, 86, 30, 19, 49, 91, 36,
       90, 77, 66, 59, 20, 23, 95, 49,  1, 12, 31, 56, 73,  8, 89, 16],
      dtype=int32)>

In [72]:
# Find the minimum
tf.reduce_min(E)

<tf.Tensor: shape=(), dtype=int32, numpy=1>

In [73]:
# Find the maximum
tf.reduce_max(E)

<tf.Tensor: shape=(), dtype=int32, numpy=98>

In [74]:
# Find the mean
tf.reduce_mean(E)

<tf.Tensor: shape=(), dtype=int32, numpy=49>

In [75]:
# Find the sum
tf.reduce_sum(E)

<tf.Tensor: shape=(), dtype=int32, numpy=2469>

You can also find the standard deviation `tf.reduce_std()` and variance `tf.reduce_variance()` of elements in a ensor uding similar methods.

### Finding the positional maximum and minimum

Say we want to find the position of the maximum value in a tensor.

This is helpful, when you want to line up your labels, like `['Green', 'Blue', 'Red']` with your prediction probabilities tensor `[0.98, 0.01, 0.01]`.

In this case, the predicted label (the one with the highest prediction probability) is `Green`. Though we won't always know the predicted label by looking at the prediction probabilities only. Thus we have the positional to help us find it.

You can do the same for minimum with the following:

`tf.argmax()` - find the max element's position
`tf.argmim()` - find the min element's position

In [76]:
# Create a tensor with 50 values between 0 and 1
F = tf.constant(np.random.random(50))
F

<tf.Tensor: shape=(50,), dtype=float64, numpy=
array([0.48551008, 0.82747985, 0.90730265, 0.34506211, 0.03348001,
       0.69003013, 0.65052136, 0.34591901, 0.85289438, 0.13253629,
       0.39117839, 0.06870841, 0.64977751, 0.19646184, 0.6061449 ,
       0.53312323, 0.25243483, 0.69895809, 0.25888549, 0.22285226,
       0.72302091, 0.19667219, 0.23853042, 0.26310526, 0.01228497,
       0.97211127, 0.09810491, 0.66451587, 0.5948059 , 0.37632583,
       0.27540981, 0.86569755, 0.11427361, 0.19401338, 0.43201752,
       0.77977753, 0.32247213, 0.60696442, 0.28119469, 0.15859664,
       0.91636374, 0.71994465, 0.74919283, 0.9307283 , 0.01533177,
       0.38774824, 0.02033448, 0.21588216, 0.98796492, 0.79500156])>

In [77]:
# Find the max element position in F
tf.argmax(F)

<tf.Tensor: shape=(), dtype=int64, numpy=48>

In [78]:
# Find the min element position in F
tf.argmin(F)

<tf.Tensor: shape=(), dtype=int64, numpy=24>

In [80]:
print(f'The maximum value of F is at position: {tf.argmax(F).numpy()}') #.numpy() > only taking the results shown in numpy
print(f'The maximum value of F is: {tf.reduce_max(F).numpy()}')
print(f'Using tf.argma() to index F, the maximum value of F is: {F[tf.argmax(F)].numpy()}') # [place index value here] aka the '48'
print(f'Now, are the two maxes the same? {F[tf.argmax(F)].numpy() == tf.reduce_max(F)}') 

The maximum value of F is at position: 48
The maximum value of F is: 0.9879649201116832
Using tf.argma() to index F, the maximum value of F is: 0.9879649201116832
Now, are the two maxes the same? True


### Squeezing a tensor (removing all single dimensions)

If you need to remove single-dimensions from a tensor (dimensions with size 1), you can use `tf.squeeze()`, which removes all dimensions with a `1` from the tensor.

In [82]:
# Create a rank 5 (5 dimensions) tensor of 50 numbers between 0 to 100
G = tf.constant(np.random.randint(0, 100, 50), shape=(1,1,1,1,50))
G.shape, G.ndim

(TensorShape([1, 1, 1, 1, 50]), 5)

In [85]:
# Squeeze tensor G (remove all 1 dimensions)
G_squeezed = tf.squeeze(G)
G_squeezed.shape, G_squeezed.ndim

(TensorShape([50]), 1)

### One hot encoding

If you have tensor indixes, and want to one-hot encode it, use `tf.one_hot()`.

Also you should specify the `depth` parameter. The level which you want to one-hot encode to.

In [86]:
# Create a list of indices
some_list = [0,1,2,3]

# One hot encode them
tf.one_hot(some_list, depth=4)

<tf.Tensor: shape=(4, 4), dtype=float32, numpy=
array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]], dtype=float32)>

You can also specify values for `on_value` and `off_value`, rather than 1 and 0.

In [87]:
# Specify custom values for on and off encoding
tf.one_hot(some_list, depth=4, on_value="We're live!", off_value="Offline")

<tf.Tensor: shape=(4, 4), dtype=string, numpy=
array([[b"We're live!", b'Offline', b'Offline', b'Offline'],
       [b'Offline', b"We're live!", b'Offline', b'Offline'],
       [b'Offline', b'Offline', b"We're live!", b'Offline'],
       [b'Offline', b'Offline', b'Offline', b"We're live!"]], dtype=object)>

### Squaring, log, square root

There are many mathematical operations you'll perform with a project, and are likely to exist.

We'll take a look at a few:
* `tf.square()` - get the square of every value in a tensor
* `tf.sqrt()` - get the square root of every value in a tensor (**Note**: the tensor has to be a float, or will return error).
* `tf.math.log()` - get the natural log of every value in a tensor (elements also need to be floats).

In [88]:
# Create a new tensor
H = tf.constant(np.arange(1,10))
H

<tf.Tensor: shape=(9,), dtype=int64, numpy=array([1, 2, 3, 4, 5, 6, 7, 8, 9])>

In [89]:
# Square it
tf.square(H)

<tf.Tensor: shape=(9,), dtype=int64, numpy=array([ 1,  4,  9, 16, 25, 36, 49, 64, 81])>

In [90]:
# Find the squareroot
tf.sqrt(H)

InvalidArgumentError: Value for attr 'T' of int64 is not in the list of allowed values: bfloat16, half, float, double, complex64, complex128
	; NodeDef: {{node Sqrt}}; Op<name=Sqrt; signature=x:T -> y:T; attr=T:type,allowed=[DT_BFLOAT16, DT_HALF, DT_FLOAT, DT_DOUBLE, DT_COMPLEX64, DT_COMPLEX128]> [Op:Sqrt] name: 

In [91]:
# Change H to float32
H = tf.cast(H, dtype=tf.float32)
H

<tf.Tensor: shape=(9,), dtype=float32, numpy=array([1., 2., 3., 4., 5., 6., 7., 8., 9.], dtype=float32)>

In [92]:
tf.sqrt(H)

<tf.Tensor: shape=(9,), dtype=float32, numpy=
array([1.       , 1.4142135, 1.7320508, 2.       , 2.236068 , 2.4494898,
       2.6457512, 2.828427 , 3.       ], dtype=float32)>

In [93]:
# Find the log (input also needs to be float)
tf.math.log(H)

<tf.Tensor: shape=(9,), dtype=float32, numpy=
array([0.       , 0.6931472, 1.0986123, 1.3862944, 1.609438 , 1.7917595,
       1.9459102, 2.0794415, 2.1972246], dtype=float32)>

### Manipulating `tf.variable()` tensors

Tensors with `tf.variable()` can be changed with the following methods:

* `.assign()` - assign a different value to a given index, in a tensor.

* `.assign_add()` - select the given index in the tensor, and add value onto it.

In [94]:
# Create a variable tensor
I = tf.Variable(np.arange(0,5))
I

<tf.Variable 'Variable:0' shape=(5,) dtype=int64, numpy=array([0, 1, 2, 3, 4])>

In [96]:
# Assign the final value with 50
I.assign([0,1,2,3,50])

<tf.Variable 'UnreadVariable' shape=(5,) dtype=int64, numpy=array([ 0,  1,  2,  3, 50])>

In [98]:
# Add 10 to every element in I
I.assign_add([10,10,10,10,10])

<tf.Variable 'UnreadVariable' shape=(5,) dtype=int64, numpy=array([10, 11, 12, 13, 60])>

### Tensors and NumPy

We've seen some examples of both NumPy arrays and Tensors interacting together, like using NumPy arrays to create tensors.

Tensors can also be converted into NumPy arrays using:

* `np.array()` - pas a tensor to convert to an ndarray (Numpy's main datatype)
* `tensor.numpy()` - call on a tensor to convert to an ndarray.

Doing this is helpful as it makes tensors iterable as well as allows us to use any of NumPy's methods on them.

In [99]:
# Create a tensor from NumPy arrays
J = tf.constant(np.array([3., 7., 10.]))
J

<tf.Tensor: shape=(3,), dtype=float64, numpy=array([ 3.,  7., 10.])>

In [100]:
# Convert tensor J to NumPy with np.array()
np.array(J), type(np.array(J))

(array([ 3.,  7., 10.]), numpy.ndarray)

In [101]:
# Conver tensor J to NumPy with .numpy()
J.numpy(), type(J.numpy())

(array([ 3.,  7., 10.]), numpy.ndarray)

By default, tensors have `dtype=float32`, where as NumPy is `dtype=float64`

This is because, neural networks (which tends to be built with TensorFlow), generally works better with less precision, preffering 32-bit, over 64-bit.

In [103]:
# Create  tensor from NumPy and from an array
numpy_J = tf.constant(np.array([3., 7., 10.])) # will be float64
tensor_J = tf.constant([3., 7., 10.]) # will be float32
numpy_J.dtype, tensor_J.dtype

(tf.float64, tf.float32)

### Using `@tf.function`

Through your TensorFlow journey, you might've come across this python function, with a decorator called `@tf.function`.

What are python decorators?

It's a design pattern, to allow for modification and extension of behaviours on a epcific function or method, without changing the actual code. They are often used for code reusability, or separation converns. Making it easy to add certain functionalities, like logging, timing, access control, etc.

**How does a decorator work?**

1. Takes another function as its input
2. Wraps it with additional functionality
3. Returns the wrapped function

Here is a code snippet example. 
```python
def sandwich_decorator(func):
    def wrapper():
        print("🍞 This is the top slice of bread.")
        func()
        print("🍞 This is the bottom slice of bread.")
    return wrapper

@sandwich_decorator
def my_filling():
    print("🥓 Here’s the delicious filling!")

my_filling()
```

And it outputs:
```
🍞 This is the top slice of bread.
🥓 Here’s the delicious filling!
🍞 This is the bottom slice of bread.
```

Pay attention to `@my_decorator` and `func()`.

`@my_decorator` - Think of it as you calling a function in python, and the variable you're pulling into `@my_decorator`, is the function `say_hello()`. You got functionception!

`func()` - This is the variable of `@my_decorator`, aka the function that was pulled into the decorator. The function will perform its action within the progression of the wrapper function.

**TLDR**: Decorators are a lot like functions, but they sandwich the fillings with bread.

Back to `@tf.function` decorator use, it turns a python function into a callable TensorFlow graph. Aka if you've written python functions, and you decorate it with `@tf.function`, when exporting the code, it will attempt to conver it into a faster version of itself, aka better performance.

In [105]:
# Create a simple function
def function(x, y):
    return x ** 2 + y

x = tf.constant(np.arange(0,10))
y = tf.constant(np.arange(10,20))
function(x, y)

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([ 10,  12,  16,  22,  30,  40,  52,  66,  82, 100])>

In [107]:
# Create same function, but decorate it with tf.function
@tf.function
def tf_function(x, y):
    return x ** 2 + y

tf_function(x, y)

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([ 10,  12,  16,  22,  30,  40,  52,  66,  82, 100])>

There is no difference between the two functions.

Much of the difference happens behind the scenes. One of the main ones, is potential code speed-ups wherever possible.

### Finding access to GPUs

We've mentions GPUs plenty of time within the notebook.

So how to check if one's available?

you can check if you have access to a GPU using `tf.config.list_physical_devices()`.


In [2]:
print(tf.config.list_physical_devices('GPU'))

[]


You can also find info regarding your GPU with `!nvidia-smi`

In [3]:
!nvidia-smi

Sun Dec 29 23:01:45 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 566.36                 Driver Version: 566.36         CUDA Version: 12.7     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                  Driver-Model | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce GTX 1060      WDDM  |   00000000:01:00.0  On |                  N/A |
| N/A   62C    P8             10W /   78W |    1406MiB /   6144MiB |     30%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

# Excercise Time:

1. Create a vector, scalar, matrix and tensor with values of your choosing using tf.constant().
2. Find the shape, rank and size of the tensors you created in 1.
3. Create two tensors containing random values between 0 and 1 with shape [5, 300].
4. Multiply the two tensors you created in 3 using matrix multiplication.
5. Multiply the two tensors you created in 3 using dot product.
6. Create a tensor with random values between 0 and 1 with shape [224, 224, 3].
7. Find the min and max values of the tensor you created in 6.
8. Created a tensor with random values of shape [1, 224, 224, 3] then squeeze it to change the shape to [224, 224, 3].
9. Create a tensor with shape [10] using your own choice of values, then find the index which has the maximum value.
10. One-hot encode the tensor you created in 9.

In [4]:
# Question 1
scalar = tf.constant(70)
scalar

<tf.Tensor: shape=(), dtype=int32, numpy=70>

In [5]:
vector = tf.constant([1, 2])
vector

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([1, 2], dtype=int32)>

In [6]:
matrix = tf.constant([[3,4],
                      [5,6]])
matrix

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[3, 4],
       [5, 6]], dtype=int32)>

In [9]:
tensor = tf.constant([[[10,20],
                       [30,40]],
                       [[11,22],
                       [33,44]]])
tensor

<tf.Tensor: shape=(2, 2, 2), dtype=int32, numpy=
array([[[10, 20],
        [30, 40]],

       [[11, 22],
        [33, 44]]], dtype=int32)>

In [11]:
# Question 2
scalar.shape, vector.shape, matrix.shape, tensor.shape

(TensorShape([]),
 TensorShape([2]),
 TensorShape([2, 2]),
 TensorShape([2, 2, 2]))

In [12]:
tf.rank(scalar), tf.rank(vector), tf.rank(matrix), tf.rank(tensor)

(<tf.Tensor: shape=(), dtype=int32, numpy=0>,
 <tf.Tensor: shape=(), dtype=int32, numpy=1>,
 <tf.Tensor: shape=(), dtype=int32, numpy=2>,
 <tf.Tensor: shape=(), dtype=int32, numpy=3>)

In [14]:
tf.size(scalar).numpy(), tf.size(vector).numpy(), tf.size(matrix).numpy(), tf.size(tensor).numpy()

(np.int32(1), np.int32(2), np.int32(4), np.int32(8))

In [21]:
# Question 3
tensor1 = tf.constant(np.random.random([5, 300]))
tensor2 = tf.constant(np.random.random([5, 300]))

tensor1, tensor2

(<tf.Tensor: shape=(5, 300), dtype=float64, numpy=
 array([[0.79996396, 0.52318446, 0.18748098, ..., 0.84164263, 0.36373581,
         0.6448981 ],
        [0.59194398, 0.82549503, 0.95528895, ..., 0.46274718, 0.20646377,
         0.21288059],
        [0.85775223, 0.32515423, 0.44863049, ..., 0.39706708, 0.22056834,
         0.34573619],
        [0.61718291, 0.08334234, 0.39282044, ..., 0.71577625, 0.583205  ,
         0.87658822],
        [0.79597198, 0.40124205, 0.86456206, ..., 0.74309565, 0.67433257,
         0.2128557 ]])>,
 <tf.Tensor: shape=(5, 300), dtype=float64, numpy=
 array([[0.29866785, 0.89096465, 0.17977047, ..., 0.78279121, 0.45629769,
         0.85754862],
        [0.82462624, 0.05751266, 0.83938867, ..., 0.76798208, 0.61105176,
         0.07967464],
        [0.97257552, 0.72587492, 0.99914081, ..., 0.07927233, 0.13682753,
         0.69810842],
        [0.58826884, 0.09078152, 0.07491041, ..., 0.4962493 , 0.18758016,
         0.01094319],
        [0.11854349, 0.46352853

In [22]:
# Question 4
result = tensor1 * tensor2
result

<tf.Tensor: shape=(5, 300), dtype=float64, numpy=
array([[0.23892352, 0.46613885, 0.03370354, ..., 0.65883046, 0.16597181,
        0.55303148],
       [0.48813253, 0.04747642, 0.80185872, ..., 0.35538154, 0.12616005,
        0.01696119],
       [0.83422882, 0.2360213 , 0.44824503, ..., 0.03147643, 0.03017982,
        0.24136134],
       [0.36306948, 0.00756594, 0.02942634, ..., 0.35520346, 0.10939769,
        0.00959267],
       [0.0943573 , 0.18598714, 0.1329056 , ..., 0.16712913, 0.231151  ,
        0.11728423]])>

In [25]:
# Question 5
tensor1 @ tf.transpose(tensor2)

<tf.Tensor: shape=(5, 5), dtype=float64, numpy=
array([[72.30981098, 75.50588583, 73.48252169, 71.86139525, 78.78838429],
       [71.54323424, 74.3869124 , 76.76940419, 74.03959575, 77.39483418],
       [72.59833209, 74.97011183, 75.3803527 , 72.16246385, 74.32319731],
       [67.21703465, 73.19693713, 66.31472193, 69.47054779, 69.22329379],
       [72.24807616, 78.28668899, 74.20517949, 73.20410387, 78.74978938]])>

In [28]:
# Question 6
tensor3 = tf.constant(np.random.random([224,224,3]))
tensor3

<tf.Tensor: shape=(224, 224, 3), dtype=float64, numpy=
array([[[0.6889541 , 0.17305598, 0.97795842],
        [0.14369852, 0.13033119, 0.91988337],
        [0.74345245, 0.67977389, 0.77476975],
        ...,
        [0.59367417, 0.99584037, 0.93725872],
        [0.87229166, 0.8970433 , 0.66983146],
        [0.78633408, 0.28153962, 0.64171979]],

       [[0.19222755, 0.74718778, 0.06603275],
        [0.82714114, 0.16598692, 0.42310519],
        [0.53803203, 0.68914195, 0.09414146],
        ...,
        [0.69968675, 0.5049401 , 0.15663633],
        [0.89616921, 0.94341304, 0.55241393],
        [0.82165558, 0.6304075 , 0.07611324]],

       [[0.30015334, 0.16433831, 0.52319256],
        [0.69017942, 0.57231043, 0.38299878],
        [0.60286582, 0.56218911, 0.9627808 ],
        ...,
        [0.8273599 , 0.16906943, 0.45210786],
        [0.20994204, 0.83433648, 0.46244244],
        [0.20675567, 0.66419667, 0.74982533]],

       ...,

       [[0.7524226 , 0.95527425, 0.44377609],
        [0.76

In [29]:
# Question 7
tf.reduce_min(tensor3), tf.reduce_max(tensor3)

(<tf.Tensor: shape=(), dtype=float64, numpy=3.099144085028094e-06>,
 <tf.Tensor: shape=(), dtype=float64, numpy=0.9999906407866549>)

In [30]:
# Question 8
tensor4 = tf.constant(np.random.random([1,224,224,3]))
tensor4

<tf.Tensor: shape=(1, 224, 224, 3), dtype=float64, numpy=
array([[[[0.00939266, 0.27975762, 0.71827679],
         [0.08894262, 0.26753247, 0.58705822],
         [0.61610311, 0.26403068, 0.71245644],
         ...,
         [0.97710813, 0.00165777, 0.45119487],
         [0.83878267, 0.43712768, 0.91458832],
         [0.32766109, 0.88423251, 0.81354076]],

        [[0.91144482, 0.02459742, 0.50308224],
         [0.20052408, 0.20453454, 0.70665027],
         [0.74788114, 0.08580635, 0.19384476],
         ...,
         [0.39945736, 0.31899759, 0.72232037],
         [0.5789252 , 0.61959607, 0.474825  ],
         [0.85404447, 0.37995542, 0.9300012 ]],

        [[0.11014093, 0.29327816, 0.54847112],
         [0.12290545, 0.07312398, 0.73941314],
         [0.7616844 , 0.51641653, 0.62406008],
         ...,
         [0.85097793, 0.79080244, 0.27823159],
         [0.27182723, 0.15493547, 0.25084947],
         [0.40124003, 0.30687909, 0.89107455]],

        ...,

        [[0.32117546, 0.50413389, 

In [31]:
# Question 9
tensor4_squeezed = tf.squeeze(tensor4)
tensor4_squeezed

<tf.Tensor: shape=(224, 224, 3), dtype=float64, numpy=
array([[[0.00939266, 0.27975762, 0.71827679],
        [0.08894262, 0.26753247, 0.58705822],
        [0.61610311, 0.26403068, 0.71245644],
        ...,
        [0.97710813, 0.00165777, 0.45119487],
        [0.83878267, 0.43712768, 0.91458832],
        [0.32766109, 0.88423251, 0.81354076]],

       [[0.91144482, 0.02459742, 0.50308224],
        [0.20052408, 0.20453454, 0.70665027],
        [0.74788114, 0.08580635, 0.19384476],
        ...,
        [0.39945736, 0.31899759, 0.72232037],
        [0.5789252 , 0.61959607, 0.474825  ],
        [0.85404447, 0.37995542, 0.9300012 ]],

       [[0.11014093, 0.29327816, 0.54847112],
        [0.12290545, 0.07312398, 0.73941314],
        [0.7616844 , 0.51641653, 0.62406008],
        ...,
        [0.85097793, 0.79080244, 0.27823159],
        [0.27182723, 0.15493547, 0.25084947],
        [0.40124003, 0.30687909, 0.89107455]],

       ...,

       [[0.32117546, 0.50413389, 0.62174468],
        [0.32

In [35]:
# Question 9
a = tf.constant([1,4,6,3,7,9,2,6,8,4])
a.shape, tf.argmax(a).numpy()

(TensorShape([10]), np.int64(5))

In [37]:
# Question 10
tf.one_hot(a, depth=10)

<tf.Tensor: shape=(10, 10), dtype=float32, numpy=
array([[0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1.],
       [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1., 0., 0., 0., 0., 0.]], dtype=float32)>